lib/multi_blas_quda.cu

```diff
@@ -701,17 +702,18 @@ namespace quda {
             tmpmajor1[count1++] = a_[count++];
         }
 
-        axpy_recurse<Functor>(tmpmajor0, x, y0, i_idx, 2 * j_idx + 0, upper);
-        axpy_recurse<Functor>(tmpmajor1, x, y1, i_idx, 2 * j_idx + 1, upper);
+        axpy_recurse<Functor>(tmpmajor0, x, y0, range_x, range(range_y.first, range_y.first + y0.size()), upper);
+        axpy_recurse<Functor>(tmpmajor1, x, y1, range_x, range(range_y.first + y0.size(), range_y.second), upper);
 
         delete[] tmpmajor;
       } else {
         // if at the bottom of recursion,
-        // return if on lower left for upper triangular,
-        // return if on upper right for lower triangular.
         if (is_valid_NXZ(x.size(), false, x[0]->Precision() < QUDA_SINGLE_PRECISION)) {
-          if (upper == 1 && j_idx < i_idx) { return; }
-          if (upper == -1 && j_idx > i_idx) { return; }
+          // since tile range is [first,second), e.g., [first,second-1], we need >= here
+          // if upper triangular and upper-right tile corner is below diagonal return
+          if (upper == 1 && range_y.first >= range_x.second) { return; }
```

Comment:
Took me a while to picture this in my head, but now that I understand it, I agree with this logic (insofar as I don't think there's a corner case testing hasn't caught yet).

Reply:
I have some ideas how to clean this up significantly, but I'll save that for a new branch.  Namely to create a `tile` object that bakes this logic into it.  From which we can query its properties, such as `isLower`, `isUpper`, etc.  Only need to implement it once and use it in this and extend it into multi-reduce, etc.